riot_pal/dut_pyshell.py

```diff
@@ -0,0 +1,258 @@
+                        what is needed (default: True)
+```
+"""
+import cmd
+from json import dumps
+import logging
+import argparse
+import serial.tools.list_ports
+try:
+    import readline
+except ImportError:
+    readline = None
+try:
+    from .dut_shell import DutShell, RESULT_SUCCESS
+except ImportError:
+    from dut_shell import DutShell, RESULT_SUCCESS
+
+
+class DutPyShell(cmd.Cmd):
+    """Command loop for the PHiLIP interface
+
+    Args:
+        port - Serial port for the PHiLIP, if None connection wizard tries to
+               connent
+        data_only - If true only data prints from command an not the whole
+                    response struct
+    """
+    prompt = 'node: '
+
+    def __init__(self, port=None, data_only=True):
+        if port is None:
+            self.dut = self._connect_wizard()
+        else:
+            self.dut = DutShell(port, parser='json')
+        self.cmd_list = self.dut.send_cmd('help')['data']
+        self.data_only = data_only
+        cmd.Cmd.__init__(self)
+
+    @staticmethod
+    def _connect_wizard():
+        print("Starting DUT python shell")
+        serial_devices = serial.tools.list_ports.comports()
+        if len(serial_devices) == 0:
+            raise ConnectionError("Could not find any available devices")
+        elif len(serial_devices) == 1:
+            print('Connected to {}'.format(serial_devices[0]))
+            return DutShell(port=serial_devices[0][0], parser='json')
+        else:
+            print('Select a serial port:')
+            for i, s_dev in enumerate(serial_devices):
+                print("{}: {}".format(i, s_dev))
+            s_num = input("Selection(number): ")
+            return DutShell(port=serial_devices[int(s_num)][0], parser='json')
+
+    def preloop(self):
+        if readline:
+            try:
+                readline.read_history_file()
+            except IOError:
+                pass
+
+    def do_send_cmd(self, arg):
+        """Read a register defined by the memory map
+
+        Usage:
+            read_reg <cmd_name> [offset] [size] [to_byte_array] [timeout]
+
+        Args:
+            cmd_name: The name of the register to read
+            offset: The number of elements to offset in an array
+            size: The number of elements to read in an array
+            to_byte_array: If True and data is bytes leave it as an array
+            timeout: Optional timeout value for command specific timeouts
+
+        """
+        try:
+            results = self.dut.send_cmd(arg)
+        except KeyError as exc:
+            print('Could not parse argument {}'.format(exc))
+        except (TypeError, ValueError, SyntaxError) as exc:
+            print(exc)
+        else:
+            self._print_func_result_success(results)
+
+    def complete_send_cmd(self, text, line, begidx, endidx):
+        """Completes arg with memory map record names"""
+        begidx = begidx
```

Comment:
is this just a dummy function?

Reply:
Nope, this allows autocomplete of functions, at least if the help is structured!

Reply:
okay